WalletWasabi/Extensions/LinqExtensions.cs

```diff
@@ -83,6 +83,26 @@ public static bool NotNullAndNotEmpty<T>(this IEnumerable<T> source)
 			return !(source is null) && source.Any();
 		}
 
+		public static IEnumerable<IEnumerable<T>> CombinationsWithoutRepetition<T>(
+			this IEnumerable<T> items,
+			int ofLength)
+		{
+			return (ofLength == 1) ?
+				items.Select(item => new[] { item }) :
+				items.SelectMany((item, i) => items.Skip(i + 1)
+												.CombinationsWithoutRepetition(ofLength - 1)
+												.Select(result => new T[] { item }.Concat(result)));
+		}
+
+		public static IEnumerable<IEnumerable<T>> CombinationsWithoutRepetition<T>(
+			this IEnumerable<T> items,
+			int ofLength,
+			int upToLength)
+		{
+			return Enumerable.Range(ofLength, Math.Max(0, upToLength - ofLength + 1))
+							.SelectMany(len => items.CombinationsWithoutRepetition(ofLength: len));
+		}
+
```

Comment:
Aren't these dangerous? Aren't these exponentially more expensive computations?

Reply:
It depends on the number of clusters. I have limited the combinations of clusters to 6 because in that way, if the user selects coins from 10 different clusters then the number of combinations is 456 (a very inexpensive operation). In my opinion, this will still work perfectly well for coins from 20 different clusters. For more than 20 clusters this will not work well; yes, I will fix that.

Reply:
Fixed. If the user selects coins from more than ten different clusters then we don't use the combinations and the algorithm works as before.